packages/client/src/links/wsLink/wsClient/requestManager.ts

```diff
@@ -0,0 +1,181 @@
+import type { AnyTRPCRouter, inferRouterError } from '@trpc/server';
+import type { Observer } from '@trpc/server/observable';
+import type {
+  TRPCClientOutgoingMessage,
+  TRPCResponseMessage,
+} from '@trpc/server/unstable-core-do-not-import';
+import type { TRPCClientError } from '../../../TRPCClientError';
+import { withResolvers } from './utils';
+
+export type TCallbacks = Observer<
+  TRPCResponseMessage<unknown, inferRouterError<AnyTRPCRouter>>,
+  TRPCClientError<AnyTRPCRouter>
+>;
+
+type MessageId = string;
+type MessageIdLike = string | number | null;
+
+/**
+ * Represents a WebSocket request managed by the RequestManager.
+ * Combines the network message, a utility promise (`end`) that mirrors the lifecycle
+ * handled by `callbacks`, and a set of state monitoring callbacks.
+ */
+interface Request {
+  message: TRPCClientOutgoingMessage;
+  end: Promise<void>;
+  callbacks: TCallbacks;
+}
+
+/**
+ * Manages WebSocket requests, tracking their lifecycle and providing utility methods
+ * for handling pending and active requests.
+ *
+ * - **Pending requests**: Requests that are queued and waiting to be sent.
+ * - **Active requests**: Requests that have been sent but are awaiting a response.
+ *   For subscriptions, multiple responses may be received until the subscription is closed.
+ */
+export class RequestManager {
+  /**
+   * Stores requests that are pending, meaning they are registered but not yet sent over the WebSocket.
+   */
+  private pendingRequests = new Array<Request & { id: MessageId }>();
+
+  /**
+   * Stores requests that are active, meaning they have been sent over the WebSocket
+   * but are still awaiting responses. For subscriptions, this includes requests
+   * that may receive multiple responses.
+   */
+  private activeRequest: Record<MessageId, Request> = {};
```

Comment:
Please change the language here

- `pending` is usually what we call things that are "in flight" - that's the language RQ uses


suggestion

- `pendingRequests` -> `outgoingRequests` / `idleRequests`
- `activeRequest` -> `pendingRequest`